lib/ts/Extendible.h

```diff
@@ -0,0 +1,584 @@
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  @section details Details
+
+////////////////////////////////////////////
+  Implements Extendible<Derived_t>:
+  * Extendible<Derived_t>::Schema
+  * Extendible<Derived_t>::FieldSchema
+  * Extendible<Derived_t>::FieldId<Access_t,Field_t>
+ */
+
+#pragma once
+#include "stdint.h"
+#include <typeinfo>
+#include <typeindex>
+#include <cstddef>
+#include <atomic>
+#include <cstring>
+#include <unordered_map>
+#include <type_traits>
+
+#include "ts/ink_memory.h"
+#include "ts/ink_assert.h"
+#include "AcidPtr.h"
+
+#ifndef ROUNDUP
+#define ROUNDUP(x, y) ((((x) + ((y)-1)) / (y)) * (y))
+#endif
+
+// used for C API
+typedef const void *FieldId_C;
+
+namespace MT // (multi-threaded)
+{
+/// used to store byte offsets to fields
+using ExtendibleOffset_t = uint16_t;
+/// all types must allow unblocking MT read access
+enum AccessEnum { ATOMIC, BIT, STATIC, ACIDPTR, DIRECT, C_API, NUM_ACCESS_TYPES };
+
+inline bool &
+areStaticsFrozen()
+{
+  static bool frozen = 0;
+  return frozen;
+}
+
+/**
+ * @brief Allows code (and Plugins) to declare member variables during system init.
+ *
+ * The size of this structure is actually zero, so it will not change the size of your derived class.
+ * But new and delete are overriden to use allocate enough bytes of the derived type + added fields.
+ * All bool's are packed to save space using the *Bit methods.
+ * This API is focused on thread safe data types that allow minimally blocked reading.
+ * This is templated so static variables are instanced per Derived type. B/c we need to have different
+ * size field sets.
+ *
+ * @tparam Derived_t - the class that you want to extend at runtime.
+ *
+ * @see test_Extendible.cc for examples
+ *
+ */
+template <typename Derived_t> struct Extendible {
+  //////////////////////////////////////////
+  // Internal classes
+  template <AccessEnum FieldAccess_e, typename Field_t> class FieldId; // field handle (strongly type)
+  struct FieldSchema;                                                  // field descriptor
+  class Schema;                                                        // memory layout, field container
+  // aliases
+  using BitFieldId = FieldId<BIT, bool>;
+
+  //////////////////////////////////////////
+  // Extendible static data
+  /// one schema instance per Derived_t to define contained fields
+  static Schema schema;
+
+  //////////////////////////////////////////
+  // Extendible member variables
+
+  // none
+  // this uses a single alloc and type erasing
+
+  //////////////////////////////////////////
+  // Extendible lifetime management
+  /** don't allow copy construct, that doesn't allow atomicity */
+  Extendible(Extendible &) = delete;
+  /** allocate a new object with additional field data */
+  void *operator new(size_t size);
+  /** construct all fields */
+  Extendible() { schema.call_construct(this_as_char_ptr()); }
```

Comment:
Why not have the method take a `void *`, in which case you wouldn't need this method, you could pass `this` directly.